src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java

```diff
@@ -485,4 +492,46 @@ else if (participation instanceof ProgrammingExerciseParticipation) {
         final var result = resultService.createNewExampleResultForSubmissionWithExampleSubmission(submissionId, isProgrammingExerciseWithFeedback);
         return new ResponseEntity<>(result, HttpStatus.CREATED);
     }
+
+    @PostMapping(value = "/exercises/{exerciseId}/external-submission-results")
+    @PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<Result> createResultForExternalSubmission(@PathVariable Long exerciseId, @RequestParam String studentLogin, @RequestBody Result result)
+            throws URISyntaxException {
+        log.debug("REST request to create Result for External Submission for Exercise : {}", exerciseId);
+
+        Exercise exercise = exerciseService.findOneWithAdditionalElements(exerciseId);
+        User user = userService.getUserWithGroupsAndAuthorities();
+        Optional<User> student = userService.getUserWithAuthoritiesByLogin(studentLogin);
+        Course course = exercise.getCourse();
+
+        if (!authCheckService.isAtLeastInstructorForExercise(exercise, user)) {
+            throw new AccessForbiddenException("You are not allowed to access this resource");
+        }
+        if (student.isEmpty() || !authCheckService.isAtLeastStudentInCourse(course, student.get())) {
+            throw new ResponseStatusException(HttpStatus.FAILED_DEPENDENCY, "No student found for " + studentLogin + " in exercise " + exerciseId);
+        }
+        if (exercise instanceof QuizExercise) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "External submissions are not supported for Quiz exercises.");
+        }
+
+        StudentParticipation participation = participationService.createParticipationWithEmptySubmission(exercise, student.get(), SubmissionType.EXTERNAL);
+        result.setParticipation(participation);
+
+        participation = (StudentParticipation) participationService.findOneWithEagerSubmissions(participation.getId());
+        Optional<Submission> optionalSubmission = participation.findLatestSubmission();
+        if (optionalSubmission.isEmpty()) {
+            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
+                    "Submission for " + studentLogin + " in exercise " + exerciseId + " was neither found nor could be created");
+        }
+        result.setSubmission(optionalSubmission.get());
+        Result existingResult = optionalSubmission.get().getResult();
+        if (existingResult != null) {
```

Comment:
this is a good check, however you have already saved some additional data previously into the database, so this check might be too late. Maybe it would be a good idea to analyze the situation in the beginning of this method and only create and save additional objects in the database in case all prerequisites are fulfilled

Reply:
If a result already exists, a submission and participation also already exist, so no additional data should have been saved at that point if I didn't make a mistake since they are both only created if they do not exist yet. But I agree that it's a safer approach to check the prerequisites in the beginning.

Reply:
I moved the check up before any database write operations.